onnxruntime/core/providers/webnn/data_transfer.cc

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/webnn/data_transfer.h"
+
+#include <emscripten.h>
+
+namespace onnxruntime {
+namespace webnn {
+
+bool DataTransfer::CanCopy(const OrtDevice& src_device, const OrtDevice& dst_device) const {
+  // Copying data between MLBuffers is not supported by WebNN.
+  return (dst_device.Type() == OrtDevice::GPU && src_device.Type() == OrtDevice::CPU) ||
+         (dst_device.Type() == OrtDevice::CPU && src_device.Type() == OrtDevice::GPU);
+}
+
+common::Status DataTransfer::CopyTensor(const Tensor& src, Tensor& dst) const {
+  size_t bytes = src.SizeInBytes();
+  if (bytes > 0) {
+    const void* src_data = src.DataRaw();
+    void* dst_data = dst.MutableDataRaw();
+
+    const auto& dst_device = dst.Location().device;
+
+    if (dst_device.Type() == OrtDevice::GPU) {
+      EM_ASM({
+        Module.jsepUploadBuffer($0, Module.HEAPU8.subarray($1, $1 + $2));
```

Comment:
This will try to upload ML buffer for all GPU tensors, including the initializers. We only need to upload the input/output buffer.

Reply:
I'm not sure how to tell the ORT that we don't want initializers in MLBuffers. Any ideas?

That said, since we don't create the MLBuffers until the first call to `[jsep][E|e]nsureBuffer` (currently only done by `dispatch` on the inputs and outputs), at least we aren't creating GPU buffers for the initializers.

Reply:
Is that possible to add flag to `WebNNExecutionProvider::GetDataTransfer()` check if current stage is session creation then return `nullptr`? e.g. maybe we can use `wnn_builder_` as flag, check if `wnn_builder_` is undefined, as I've already set it to `undefined` after `WebNNExecutionProvider::Compile` done.

> That said, since we don't create the MLBuffers until the first call to [jsep][E|e]nsureBuffer (currently only done by dispatch on the inputs and outputs), at least we aren't creating GPU buffers for the initializers.

This would call `jsepUploadBuffer` for every initializer, although we don't create the ml buffers for them, I am not sure how much overhead it would cause.

Reply:
> Is that possible to add flag to WebNNExecutionProvider::GetDataTransfer() check if current stage is session creation then return nullptr? e.g. maybe we can use wnn_builder_ as flag, check if wnn_builder_ is undefined, as I've already set it to undefined after WebNNExecutionProvider::Compile done.

It doesn't work, as `WebNNExecutionProvider::GetDataTransfer()` only be called once.

Reply:
AFAIK, for CudaEP/JSEP, ONNX Runtime uses a component called graph partitioner to decide which EP to run each node, and then determine which location for the input(s)/output(s) of the nodes (initializers/intermediate tensors) to place. If any initializer is considered to be used on GPU, a device copy will be performed.

I think WebNN EP works in a different way (it uses a graph builder instead of running every kernel in ORT's graph executor), so not sure how copy of initializers are controlled.

Reply:
> This will try to upload ML buffer for all GPU tensors, including the initializers. We only need to upload the input/output buffer.

In the long run, WG is considering using MLBuffer for constants / initializers. We may want to add a TODO and link to WG's MLBuffer discussion.

> That said, since we don't create the MLBuffers until the first call to [jsep][E|e]nsureBuffer (currently only done by dispatch on the inputs and outputs), at least we aren't creating GPU buffers for the initializer

At the current stage, I think if we can ensure no extra MLBuffers (already done) and ArrayBuffers (See my other comment for upload) being made for initializers, it should be fine.

Reply:
Thanks @fs-eire, I look into the session creation code, the initializers copy happens at https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/framework/session_state_utils.cc#L164, looks like it's difficult to get EP info at this stage to not copy the initializers for WebNN EP.

@egalli, I am thinking of if it's possible to add a flag from TS side, e.g. `wasm.shouldTransferMLBuffer` whatever,  similar to the `wasm.currentContext` you set in the `wasm-core-impl.ts`, then reset this flag after session created. 

With this flag we are able to distinguish if we should allocate / transfer ML buffer in `allocator.cc` and `data_transfer.cc`.

Reply:
Added `wasm.shouldTransferMLBuffer`, we no longer create MLBuffers or copies for initializers.